ckan/templates/user/list.html

```diff
@@ -16,7 +16,7 @@ <h1 class="page-heading">
         <ul class="user-list">
           {% block users_list_inner %}
             {% for user in page.items %}
-              <li>{{ h.linked_user(user['name'], maxlength=20) }}</li>
+              <li>{{ h.linked_user(user[0].name, maxlength=20) }}</li>
```

Comment:
what's going on here?

Reply:
`Session.query(User, User.name, User.id)` returns an iterator of items, that behaves in a special way.

if you write a loop like this:
```python
for item in query:
   assert item[0] is user 
   assert item[1] is user_name
   assert item[2] is user_id
```
it behaves in the expected way, allowing you to extract separate parts of the query by position. This is not changed

But if you write the following:
```
for item in query:
   assert item.email is user_email
```
where this `email` property will be taken from? Not quite expected(for me), but sqlalchemy just transforms it like the following: `item.email -> item[0].email`. And you'll get an email from the first item that was queried, from the `User` object.

In 2.0 this behavior was changed. Now you have to explicitly specify the source of an attribute. i.e, `user.email` for the given query no longer allowed.


In this case, we are passing to templates not a  simple`query(User)` but something [much more monstrous](https://github.com/ckan/ckan/blob/master/ckan/logic/action/get.py#L847-L861). That's why we have to specify the index of item, from which we are extracting the name.

Reply:
instead of passing the monstrous query object to the templates should we pass a simple list of dicts so that the template can remain unchanged and avoid the mysterious `[0].name`-style notation?

Reply:
Nope, it's not that simple.

On the `user.index` endpoint, we ask user_list for [a query object](https://github.com/ckan/ckan/blob/master/ckan/views/user.py#L90) instead of a list of dictionaries. It's already a bad decision, but I have an idea, why it's done in this way. [user_list does not support pagination](https://github.com/ckan/ckan/blob/master/ckan/logic/action/get.py#L917-L925), and to avoid pulling and dictizing **all** users, the `user.index` takes the query and passes it to `Page` object. Which in turn can compute the number of pages and fetch only the required range of users`

Adding `limit`/`offset` to the `user_list` action won't help much. This action [returns list](https://github.com/ckan/ckan/blob/master/ckan/logic/action/get.py#L925) and we won't know how many users are there, so the pagination widget will not work. It would be cool to have a result of `user_list` similar to `package_search`, with a `count` attribute. But it's a breaking change, so it's not an option for now. It's possible to solve this by making two requests: one returns a list of dictionaries restricted by new `limit`/`offset` parameters and one returns a query that we'll use to get a total number of items and build a pagination widget. But, as a result, we'll just switch from non-readable to non-efficient solution. 
So changing the API is a matter of further discussion and a separate PR.

Here, in order to preserve simplicity in the template, I can suggest using [`all_fields=False`](https://github.com/ckan/ckan/blob/master/ckan/logic/action/get.py#L917)(I already updated the PR). In this way, instead of `Query[tuple[User, str, str, str ...]]` we'll get the list of named tuples with `name` attribute in the template. I.e `user.name` will work just as it did before. As we are using only the name in the template in order to build a link to the user profile, it definitely solves the problem at the core.
The only problem that may appear: if any extension is actually relying on this monstrous object, it stops working. But I'm sure it's an acceptable risk because this structure is quite unusual for CKAN and relying on it means relying on CKAN's internal implementation details, which is never a good idea. And we still have username, so extensions should be able to get enough user details [just as we are doing in core template](https://github.com/ckan/ckan/blob/master/ckan/templates/user/list.html#L19)